File: packages/daemon/src/notifications/notifier.ts
import { spawn } from 'node:child_process'

export interface Notification {
  title: string
  message: string
  sound: boolean
}

/** Fires an OS-native notification. Per-OS implementations sit behind this interface. */
export interface Notifier {
  notify(notification: Notification): void
}

/** macOS notifications via `osascript`. Fire-and-forget; never crashes the daemon. */
export class MacNotifier implements Notifier {
  notify({ title, message, sound }: Notification): void {
    const script = `display notification ${quote(message)} with title ${quote(title)}${
      sound ? ' sound name "Ping"' : ''
    }`
    const child = spawn('osascript', ['-e', script], { stdio: 'ignore' })
    child.on('error', () => {})
  }
}

/** No-op notifier for unsupported platforms and tests. */
export class NoopNotifier implements Notifier {
  notify(): void {}
}

export function createNotifier(platform: NodeJS.Platform = process.platform): Notifier {
  return platform === 'darwin' ? new MacNotifier() : new NoopNotifier()
}

function quote(value: string): string {
  return `"${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`
}

File: crates/daemon/src/notifications/notifier.rs
use std::process::{Command, Stdio};
use std::sync::Arc;

/// A single OS notification request.
pub struct Notification {
    pub title: String,
    pub message: String,
    pub sound: bool,
}

/// Fires an OS-native notification. Per-OS implementations sit behind this trait.
pub trait Notifier: Send + Sync {
    fn notify(&self, notification: Notification);
}

/// macOS notifications via `osascript`. Fire-and-forget; never crashes the daemon.
pub struct MacNotifier;

impl Notifier for MacNotifier {
    fn notify(&self, notification: Notification) {
        let sound = if notification.sound {
            " sound name \"Ping\""
        } else {
            ""
        };
        let script = format!(
            "display notification {} with title {}{sound}",
            quote(&notification.message),
            quote(&notification.title),
        );
        let _ = Command::new("osascript")
            .args(["-e", &script])
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn();
    }
}

/// No-op notifier for unsupported platforms and tests.
pub struct NoopNotifier;

impl Notifier for NoopNotifier {
    fn notify(&self, _notification: Notification) {}
}

pub fn create_notifier() -> Arc<dyn Notifier> {
    if cfg!(target_os = "macos") {
        Arc::new(MacNotifier)
    } else {
        Arc::new(NoopNotifier)
    }
}

fn quote(value: &str) -> String {
    format!("\"{}\"", value.replace('\\', "\\\\").replace('"', "\\\""))
}
